packages/peregrine/lib/talons/SearchPage/useSearchPage.js

```diff
@@ -52,28 +75,66 @@ export const useSearchPage = props => {
     }, []);
     /* eslint-enable react-hooks/exhaustive-deps */
 
+    // Get "allowed" filters by intersection of schema and aggregations
+    const { data: introspectionData, error: introspectionError } = useQuery(
+        FilterIntrospectionQuery
+    );
+
+    useEffect(() => {
+        if (introspectionError) {
+            console.error(introspectionError);
+        }
+    }, [introspectionError]);
+
+    // Create a type map we can reference later to ensure we pass valid args
+    // to the graphql query.
+    // For example: { category_id: 'FilterEqualTypeInput', price: 'FilterRangeTypeInput' }
+    const filterTypeMap = useMemo(() => {
+        const typeMap = new Map();
+        if (introspectionData) {
+            introspectionData.__type.inputFields.forEach(({ name, type }) => {
+                typeMap.set(name, type.name);
+            });
+        }
+        return typeMap;
+    }, [introspectionData]);
+
     const pageControl = {
         currentPage,
         setPage: setCurrentPage,
         totalPages
     };
 
-    let apolloQueryVariable = {
-        currentPage: Number(currentPage),
-        inputText,
-        pageSize: PAGE_SIZE
-    };
+    const [runQuery, { loading, error, data }] = useLazyQuery(query);
 
-    if (categoryId) {
-        apolloQueryVariable = {
-            ...apolloQueryVariable,
-            categoryId
-        };
-    }
-
-    const { loading, error, data } = useQuery(query, {
-        variables: apolloQueryVariable
-    });
+    useEffect(() => {
```

Comment:
This entire effect is basically a copy pasta from the category page effect. A lot of the code changes in this file are. I could extract a good amount of this into a custom hook that would only require a few inputs but I think it's more important to get feedback on the approach before optimizing.

Reply:
I don't think it's too early to create a generic hook in `peregrine/lib/hooks` called `useFilterIntrospection`, as long as all it does is the work from L54-L76:

- run the query, which it receives as a prop: `query`
- log an error, if present
- return a memoized `filterTypeMap`

It probably wouldn't even need to return an object; it could return the map directly.

Reply:
If I was to make a special hook who would own the query? Would all components that use the hook need to know to pass that specific query? Seems like a good case for peregrine owning `gql` as a dependency 😆

Reply:
Yes, components in `venia-ui` would import/define the query and pass it to the `peregrine` hook. If we ever decide to move our queries into `peregrine`, it'll be easy to streamline hooks & talons by putting them as default arguments, while still leaving them open to custom queries. IMO.

Reply:
In this specific case the `useFilterIntrospection` hook would have logic directly tied to the graphql fields, which just makes every component required to pass the _same_ query. In that case why not just have the query live in the hook?


This is the code that is used to generate the type map. It checks `__type.inputFields.name`:

```js
    const filterTypeMap = useMemo(() => {
        const typeMap = new Map();
        if (introspectionData) {
            introspectionData.__type.inputFields.forEach(({ name, type }) => {
                typeMap.set(name, type.name);
            });
        }
        return typeMap;
    }, [introspectionData]);
```